nano/node/distributed_work.cpp

```diff
@@ -448,6 +454,27 @@ void nano::distributed_work_factory::cleanup_finished ()
 	}
 }
 
+void nano::distributed_work_factory::stop ()
+{
+	if (!stopped.exchange (true))
+	{
+		// Cancel any ongoing work
+		std::unordered_set<nano::root> roots_l;
+		nano::unique_lock<std::mutex> lock_l (mutex);
+		for (auto const & item_l : items)
+		{
+			roots_l.insert (item_l.first);
+		}
+		lock_l.unlock ();
+		for (auto const & root_l : roots_l)
+		{
+			cancel (root_l, true);
+		}
+		lock_l.lock ();
+		items.clear ();
+	}
```

Comment:
I wonder if this can be simplified using `swap` (we use that pattern a few places). The first loop and the clear would be gone. Or do you need `items` to be alive during the cancellation?

Reply:
Yeah it needs to persist, otherwise `cancel ()` won't find the root in items. An alternative here would be to replicate `cancel ()` in `stop ()`  directly, but that has code redundancy. WDYT?